File: frontend/src/utils/export-filter.ts
import type { CatalogCollection, CatalogEntity, ExportFilter, PageResponse } from '@/types/api'

export interface CatalogExportFilters {
  title: string
  author: string
  organization: string
  publicationYear?: number
  achievementType: string
  sourceCode: string
  venue: string
  topic: string
}

export type CatalogEntityResolver = (
  collection: CatalogCollection,
  name: string,
) => Promise<PageResponse<CatalogEntity>>

export class ExportFilterResolutionError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'ExportFilterResolutionError'
  }
}

const entityLabels: Record<CatalogCollection, string> = {
  authors: '作者',
  organizations: '机构',
  venues: '期刊',
  topics: '主题',
}

export async function resolveExportFilter(
  filters: CatalogExportFilters,
  resolveEntity: CatalogEntityResolver,
): Promise<ExportFilter> {
  const sourceType = resolveSourceType(filters.sourceCode)
  const [authorId, organizationId, venueId, topicId] = await Promise.all([
    resolveEntityId('authors', filters.author, resolveEntity),
    resolveEntityId('organizations', filters.organization, resolveEntity),
    resolveEntityId('venues', filters.venue, resolveEntity),
    resolveEntityId('topics', filters.topic, resolveEntity),
  ])

  return compact({
    title: normalized(filters.title),
    authorId,
    organizationId,
    publicationYearFrom: filters.publicationYear,
    publicationYearTo: filters.publicationYear,
    achievementType: normalized(filters.achievementType),
    sourceType,
    venueId,
    topicId,
  })
}

async function resolveEntityId(
  collection: CatalogCollection,
  value: string,
  resolveEntity: CatalogEntityResolver,
): Promise<number | undefined> {
  const name = normalized(value)
  if (!name) return undefined

  const page = await resolveEntity(collection, name)
  if (page.totalElements !== 1 || page.items.length !== 1) {
    const label = entityLabels[collection]
    throw new ExportFilterResolutionError(
      `${label}“${name}”匹配到 ${page.totalElements} 条记录，请调整为唯一名称后再导出`,
    )
  }
  return page.items[0]?.id
}

function resolveSourceType(value: string): ExportFilter['sourceType'] {
  const sourceType = normalized(value)?.toUpperCase()
  if (!sourceType) return undefined
  if (sourceType !== 'OPENALEX' && sourceType !== 'CROSSREF') {
    throw new ExportFilterResolutionError('来源代码仅支持 OPENALEX 或 CROSSREF')
  }
  return sourceType
}

function normalized(value: string): string | undefined {
  return value.trim() || undefined
}

function compact(filter: ExportFilter): ExportFilter {
  return Object.fromEntries(
    Object.entries(filter).filter(([, value]) => value !== undefined),
  ) as ExportFilter
}

File: systems/crawler/backend/src/main/java/com/aacv/system/export/domain/ExportFilter.java
package com.aacv.system.export.domain;

import com.aacv.system.source.domain.SourceType;

public record ExportFilter(
        String title,
        Long authorId,
        Long organizationId,
        Integer publicationYearFrom,
        Integer publicationYearTo,
        String achievementType,
        SourceType sourceType,
        Long venueId,
        Long topicId) {

    public ExportFilter {
        title = normalize(title, 512, "题名条件长度超出限制");
        achievementType = normalize(achievementType, 64, "成果类型长度超出限制");
        if (publicationYearFrom != null && (publicationYearFrom < 1000 || publicationYearFrom > 9999)
                || publicationYearTo != null && (publicationYearTo < 1000 || publicationYearTo > 9999)) {
            throw new IllegalArgumentException("发表年份无效");
        }
        if (publicationYearFrom != null && publicationYearTo != null
                && publicationYearFrom > publicationYearTo) {
            throw new IllegalArgumentException("起始年份不能晚于结束年份");
        }
        if (invalidId(authorId) || invalidId(organizationId) || invalidId(venueId) || invalidId(topicId)) {
            throw new IllegalArgumentException("导出筛选实体ID无效");
        }
    }

    private static boolean invalidId(Long value) {
        return value != null && value < 1;
    }

    private static String normalize(String value, int maxLength, String message) {
        if (value == null || value.isBlank()) {
            return null;
        }
        String normalized = value.trim();
        if (normalized.length() > maxLength) {
            throw new IllegalArgumentException(message);
        }
        return normalized;
    }
}
